src/main/java/uk/ac/cam/alpha2017/airqualityradar/backend/data/HistoricalDataProvider.java

```diff
@@ -29,15 +31,58 @@ public HistoricalDataProvider(AzureTableConnector connector, DataPointConverter
      *
      * @param fromCalendar The date and time to start return data points from
-     * @param toCalendar The date and time that we will not return data points after
+     * @param toCalendar   The date and time that we will not return data points after
      * @return The data points from the historical data per calendar per location
      */
     public List<DataPoint> getDataPoints(Calendar fromCalendar, Calendar toCalendar) throws StorageException, InvalidKeyException, URISyntaxException, TableDoesNotExistException {
-        Iterator<AirDataEntity> resultIterator = connector.getEntitiesBetweenCalendars(TABLE_NAME, fromCalendar, toCalendar);
-        List<DataPoint> dataPoints = new ArrayList<>();
+        ArrayList<AirDataEntity> airDataResultList = new ArrayList<>();
+        ArrayList<WeatherDataEntity> weatherDataResultList = new ArrayList<>();
 
-        resultIterator.forEachRemaining(entity -> dataPoints.add(dataPointConverter.convertAirAndWeatherToDataPoint(entity, null)));
+        //Convert iterators to list using Java8 syntactic sugar & lambdas <3
+        connector.getEntitiesBetweenCalendars(HISTORICAL_AIR_DATA_TABLE_NAME, fromCalendar, toCalendar, AirDataEntity.class).forEachRemaining(airDataResultList::add);
+        connector.getEntitiesBetweenCalendars(HISTORICAL_WEATHER_DATA_TABLE_NAME, fromCalendar, toCalendar, WeatherDataEntity.class).forEachRemaining(weatherDataResultList::add);
 
+        //Will error if no weatherDataResults retrieved but I think that's appropriate (otherwise mapDataPoints would error)
+        if (weatherDataResultList.size() == 0) throw new IllegalStateException("Weather data is empty");
+        return mapDataPoints(weatherDataResultList, airDataResultList);
+    }
+
+    /**
+     * Maps historical air quality data with historical weather data (both retrieved from Azure)
+     *
+     * @param weatherDataResultList A list of historical weather data retrived from Azure. ArrayList required for fast sorting.
+     * @param airDataResultList A list of historical air quality data retrived from Azure. ArrayList required for fast sorting.
+     * @return The data points including both weather and air quality from the historical data per calendar per location
+     */
+    private ArrayList<DataPoint> mapDataPoints(List<WeatherDataEntity> weatherDataResultList, List<AirDataEntity> airDataResultList) {
+        //Sort lists
+        Collections.sort(airDataResultList);
+        Collections.sort(weatherDataResultList);
+
+        //Setup
+        int weatherIndex = 0;
+        WeatherDataEntity currentWeatherEntity = weatherDataResultList.get(weatherIndex);
+        ArrayList<DataPoint> dataPoints = new ArrayList<>(weatherDataResultList.size());
+
+        for (AirDataEntity airDataEntity : airDataResultList) {
+            //if weatherDataResultList has next
+            if ((weatherIndex + 1) < weatherDataResultList.size()) {
```

Comment:
I think getting an iterator from a list should be quite fast — although we might be able to sort the results in the Azure query so we don't have to here